template.rb

```diff
@@ -1,9 +1,69 @@
 require 'shellwords'
 
+# Set Thor::Actions source path for looking up the files
+def source_paths
+  @source_paths
+end
+
+# Prepend the required paths to the source paths to make the template files in those paths available
+def use_source_path(source_path)
+  @source_paths.unshift(source_path)
+end
+
+def remote_repository
+  require 'tmpdir'
+  tempdir = Dir.mktmpdir('rails-templates')
+  at_exit { FileUtils.remove_entry(tempdir) }
+
+  git clone: [
+    '--quiet',
+    'https://github.com/nimblehq/rails-templates.git',
+    tempdir
+  ].map(&:shellescape).join(' ')
+
+  if (branch = __FILE__[%r{rails-templates/(.+)/template.rb}, 1])
+    Dir.chdir(tempdir) { git checkout: branch }
+  end
+
+  tempdir
+end
+
+def delete_test_folder
+  FileUtils.rm_rf('test')
+end
+
+def print_error_message
+  puts <<~EOT
+    #{'=' * 80}
+
+    There are some errors when templating the application, Please fix them manually:
+
+    #{@template_errors}
+    #{'=' * 80}
+  EOT
+end
+
+# Setup the template root path
+# If the template file is the url, clone the repo to the tmp directory
+template_root = __FILE__ =~ %r{\Ahttps?://} ? remote_repository : __dir__
```

Comment:
Which `tmp` directory? 🤔 Why is this needed?

Reply:
This is for when we use the template from the url like `rails new my-app -m https://raw.github...`. We need to clone the repo into local before using, so it will be clone to a tmp directory created by `Dir.mktmpdir`

![Screen Shot 2563-03-18 at 14 34 06](https://user-images.githubusercontent.com/6965195/76936511-a3bcfe80-6925-11ea-8c94-17fab54478c0.png)